physcraper/__init__.py

```diff
@@ -980,48 +982,17 @@ def add_otu(self, gb_id, ids_obj):
             debug(gb_id)
         otu_id = "otuPS{}".format(self.ps_otu)
         self.ps_otu += 1
-        ncbi_id = None
-        tax_name = None
         ott_id = None
-        if gb_id[:6] == "unpubl":  # There may not be ncbi id, because they aren't published
-            tax_name = self.gb_dict[gb_id]["^ot:ottTaxonName"]
-            ncbi_id = self.gb_dict[gb_id]["^ncbi:taxon"]
-            ott_id = self.gb_dict[gb_id]["^ot:ottId"]
-            if tax_name is None:
-                tax_name = self.gb_dict[gb_id][u'^user:TaxonName']
-            if ncbi_id is None: 
-                # debug(tax_name.split(" ")[0])
-                tax_lin_name = tax_name.split(" ")[0]
-                tax_lin_name = tax_lin_name.split("_")[0]
-                # debug(tax_lin_name)
-                ncbi_id = ids_obj.ncbi_parser.get_id_from_name(tax_lin_name)
-        elif len(gb_id.split(".")) >= 2:  # used to figure out if gb_id is from Genbank
-            if gb_id in self.gb_dict.keys() and "staxids" in self.gb_dict[gb_id].keys():
-                tax_name = self.gb_dict[gb_id]["sscinames"]
-                ncbi_id = self.gb_dict[gb_id]["staxids"]
-            else:  # all web blast results
-                tax_name = ids_obj.find_name(acc=gb_id)
-                if tax_name is None:
-                    sys.stderr.write("no species name returned for {}".format(gb_id))
-                ncbi_id = ids_obj.map_acc_ncbi(gb_id)
-        else:
-            sys.stderr.write("Something is wrong, I cannot add a new seq which has no gb_id or is not unpublished.")
-            exit(-1)
-        if ncbi_id is None:
-            # debug("ncbi_id is none")
-            if ids_obj.otu_rank is not None: 
-                ncbi_id = ids_obj.get_rank_info_from_web(tax_name)
-            #     ncbi_id = ids_obj.otu_rank[tax_name]["taxon id"]
-            # else:
-            #     ncbi_id = ids_obj.ncbi_parser.get_id_from_name(tax_name)
-            ids_obj.acc_ncbi_dict[gb_id] = ncbi_id
-            ids_obj.ncbiid_to_spn[ncbi_id] = tax_name
-            ids_obj.spn_to_ncbiid[tax_name] = ncbi_id
+        debug("trying to add an otu with accesion {}".format(gb_id))
+        ncbi_id, tax_name = get_tax_info_from_acc(gb_id, self, ids_obj)
+        debug("ADD OTU: accession {} ncbi_id {}, taxon_name {}".format(gb_id, ncbi_id, tax_name))
+        if ncbi_id == None:
+            return None
```

Comment:
If we want to be able to add unpublished sequences, every sequence needs some form of numerical identifier to make it work with the FilterBlast class. I would suggest that if there is no ncbi id we need to generate an 'artificial' id.

Reply:
Good to know. But our unpublished sequences should get ncbi_taxon id's in line 987, right?  If for some reason some sequences don't have taxon ids (like a new taxon that isn't in ncbi yet) I think our options are: a) set the ncbi_id to 0 (which I think results in it getting discarded in the filter blast class?) b) set the ncbi_id to None when we get to a sequence in the filter blast class, check if it has an ncbi_id  and discard it right away if not or c) some other approach. Do you have a preference among these?

Reply:
I think a good approach would be to generate an artifical physcraper internal numerical identifier for any kind of OTU. So, that we can have a dictionary with the artifical id (similar to the otu_labels) as key and then a nested dictionary with the different available identifiers and names from open tree and ncbi.

Reply:
Currently every input tip will have an ncbI_id (may be a higher taxon if spp is unknown), so this shouldn't be an issue.